Explain @ConfigGroup and @ConfigRoot in more details

sarxos opened the pull request:
Hello.

This commit is in regards to our discussion on Zulip. See https://quarkusio.zulipchat.com/#narrow/stream/187030-users/topic/Extensions for more details.

My concern was in regards to the following statement from the "2.4.6. Configuration Example" in Extensions tutorial here https://quarkus.io/guides/extension-authors-guide.

> 2. The @ConfigRoot annotation indicates that this object is a configuration root group, whose property names will have a parent only of quarkus.. In this case the properties within the group will begin with quarkus.log.*.

Where the code sample looks like this:

```java
@ConfigRoot(phase = ConfigPhase.RUN_TIME)
public class LogConfiguration {

    // ...

    /**
     * Configuration properties for the logging file handler.
     */
    FileConfig file;
}
```

And where point 2 states:

> ... indicates that this object is a configuration root group, whose property names will have a parent only of quarkus.. In this case the properties within the group will begin with quarkus.log.*.

But where is this indication? How this class is linked to `quarkus.log.*`? Is there some declaration missing?

----

This PR is to (hopefully) make this paragraph more clear and to document details you guys provided to me in the Zulip chat, since these details are not explained anywhere else.

sarxos committed:
Explain @ConfigGroup and @ConfigRoot in more details

Address code review comments in regards to log file name

dmlloyd commented on docs/src/main/asciidoc/extension-authors-guide.adoc:
```diff
@@ -494,15 +494,15 @@ public class FileConfig {
+     * The file name where logs will be stored.
      */
-    @ConfigItem(defaultValue = "quarkus.log")
+    @ConfigItem(defaultValue = "application.log")
```
This change should be accompanied by an update to `.gitignore` to change `quarkus.log` to `application.log`, else we'll see it show up accidentally in commits before long.

dmlloyd replied:
I also see a reference to `quarkus.log` in `test-framework/common/src/main/java/io/quarkus/test/common/PropertyTestUtil.java`.  Better make sure that won't be affected.

gsmet replied:
Should we really change the default log filename? I don't see the new one as an improvement.

Let's remove this change from this PR, it should be discussed separately as mentioned by @dmlloyd .

sarxos replied:
Hi @dmlloyd, @gsmet. I do not agree. Please note that this modification does not change any logic linked to logging, nor default logs placement nor any file names connected to logging mechanisms. This change is embedded in context of this specific example, an example demonstrating how configuration can be structured into classes, and should be treated as such (as example, not as a reference where logs are located and in which files). I modified this line because file name used here, namely `quarkus.log`, is very confusing in regards to the matter of this example, where users need to understand how it coincides with the `quarkus.log` namespace (not file name) used in configuration. So not to confuse `quarkus.log` namespace with `quarkus.log` file I changed it to `application.log`.

sarxos replied:
So what I would suggest to have your comments addressed and keep the example simple and straightforward, is to change `FileConfig.java` by adding:

```java
public static final String DEFAULT_LOG_FILE_NAME = "quarkus.log";
```

And use it there:

```java
@ConfigItem(defaultValue = DEFAULT_LOG_FILE_NAME)
File path;
```

And in `PropertyTestUtil.java` by changing:

```java
public static String getLogFileLocation() {
    return getLogFileLocation("quarkus.log");
}
```

To:

```java
public static String getLogFileLocation() {
    return getLogFileLocation(FileConfig.DEFAULT_LOG_FILE_NAME);
}
```

Same static reference will be used in example we are discussing and keep it non confusing. I hope this is ok.

sarxos replied:
Please see https://github.com/quarkusio/quarkus/pull/2589/commits/5ed96c05b1e317e973a3a3eeea0ed44a4a596ed3

dmlloyd replied:
Maybe using an existing system like logging in this doc was a mistake. :)

We'll always be trying to keep it in sync, and there will be questions when it isn't.

dmlloyd replied:
Also it makes the PR difficult to review...

sarxos replied:
@dmlloyd Yes, it would be better if we came up with a better example in the past (not an exact copy of internal logging system configuration), but since it's already there it would be a waste of time to implement something new. IMO this example is sufficient and we should simply not try to keep it in sync since this example is only to demonstrate how configuration can be composed into structure, not how logging can, or should be configured. Moreover, it's already out-of-sync because e.g. `RotationConfig` is missing in example, where it's used in the real system.

dmlloyd replied:
All the more reason why it's a bad example.  And it's clearly not a waste of time to change it at all, because you would invest time and the result would be something clearer (forever).  Saying it's a waste of time to fix this is like saying that it's a waste of time to ever improve anything, so long as it meets a bare minimum standard.  It's OK to have high standards!

dmlloyd commented on docs/src/main/asciidoc/extension-authors-guide.adoc:
```diff
@@ -482,7 +482,7 @@ import java.util.logging.Level;
 
     /**
-     * Enable file logging.
+     * Enable logging to file.
```
"to *a* file" maybe?

sarxos replied:
Done.

dmlloyd commented on docs/src/main/asciidoc/extension-authors-guide.adoc:
```diff
@@ -494,15 +494,15 @@ public class FileConfig {
 
     /**
-     * The file log level.
+     * The log level.
```
Why remove the qualification here?  There are several different log level configuration items, so it's good to clarify what scope each applies to.

sarxos replied:
@dmlloyd What you said is true in a broader view. There are more log level items in a configuration, but in this example there is only one, a `file`, so due to low level of details explained in this paragraph, having this qualifier was confusing in regards to the matter of this specific example (is this about a file `path`, or a `file` namespace). I can change this back, but only after adding more details to the description, i.e. "The level of logs to be written into the file". Please let me know if you accept this.

sarxos replied:
Please see https://github.com/quarkusio/quarkus/pull/2589/commits/5ed96c05b1e317e973a3a3eeea0ed44a4a596ed3

dmlloyd reviewed (dismissed):
Looks mostly OK with a couple small issues.

dmlloyd commented:
In fact, changing the default log file name is pretty significant and shouldn't be "hidden" as a minor doc change.

gsmet commented:
@sarxos could you take our comments into account and push an update? Thanks!

gsmet commented on docs/src/main/asciidoc/extension-authors-guide.adoc:
```diff
@@ -531,20 +531,37 @@ public class LoggingProcessor {
 quarkus.log.file.path=/tmp/debug.log
 ----
 
+Since `format` is not defined in these properties, a default value from `@ConfigItem` will be used instead.
```
s/a default value/the default value/

sarxos replied:
Will do.

gsmet commented on docs/src/main/asciidoc/extension-authors-guide.adoc:
```diff
@@ -531,20 +531,37 @@ public class LoggingProcessor {
-<3> Here the `LoggingProcessor` injects a `LogConfiguration` instance automatically by detecting the `@ConfigRoot` annotation.
+<2> The `@ConfigRoot` annotation indicates that this object is a configuration root group, in this case one which
+corresponds to a `log` segment. A class name is used to link configuration root group with the segment from a
+property name. This class name is `LogConfiguration` so the `Configuration` part is stripped off, the remaining `Log`
```
s/so the/: the/
There's no consequence here.

sarxos replied:
Will change.

gsmet commented on docs/src/main/asciidoc/extension-authors-guide.adoc:
```diff
@@ -531,20 +531,37 @@ public class LoggingProcessor {
+corresponds to a `log` segment. A class name is used to link configuration root group with the segment from a
+property name. This class name is `LogConfiguration` so the `Configuration` part is stripped off, the remaining `Log`
+is lowercased and becomes a `log`. Since all `@ConfigRoot` annotated classes uses `quarkus` as a prefix, this finally
+becomes `quarkus.log` and represents a properties which names begins with `quarkus.log.*`.
```
```suggestion
becomes `quarkus.log` and represents the properties which names begin with `quarkus.log.*`.
```

sarxos replied:
Will change.

gsmet reviewed (changes requested):
I added a couple of minor comments.

sarxos commented:
@dmlloyd @gsmet Thank you for review! Before I address your comments I would like to discuss them. Especially the log matter because I changed this intentionally to make this example less confusing. Let's discuss under the comments.

sarxos commented:
@gsmet @dmlloyd I hope I addressed all of your comments. Can we do next code review iteration and merge this PR if there are no more issues here?

gsmet commented on core/runtime/src/main/java/io/quarkus/runtime/logging/FileConfig.java:
```diff
@@ -48,7 +53,7 @@
     /**
      * The file logging log level
```
OK, so this is not related to your patch but this comment is obviously incorrect. Let's fix it?

sarxos replied:
Done.

gsmet commented on docs/src/main/asciidoc/extension-authors-guide.adoc:
```diff
@@ -494,15 +494,15 @@ public class FileConfig {
 
     /**
-     * The file logging log level.
+     * The name of file where logs will be stored.
```
Ah, in fact you spotted it, but you updated the javadoc in the comments of this documentation but not in the original source file :).

Updating the source file is definitely more important.

sarxos replied:
Done.

gsmet commented on docs/src/main/asciidoc/extension-authors-guide.adoc:
```diff
@@ -494,15 +494,15 @@ public class FileConfig {
+     * The name of file where logs will be stored.
      */
-    @ConfigItem(defaultValue = "quarkus.log")
+    @ConfigItem(defaultValue = DEFAULT_LOG_FILE_NAME)
```
The constant is not defined here.

sarxos replied:
@gsmet I didn't want to confuse reader readers with `quarkus.log` file name vs `quarkus.log` namespace and thus I didn't include it here on purpose. I thought that having a static placeholder will be enough to keep it simple and readable but not confusing. IMO, since this code is just a sample used for a demonstration purpose and to explain how configuration structures and nesting can be done, it does not have to be as strict as the runnable code. Can we leave this as it is? I would like to avoid having any mention of `quarkus.log` in terms other than a namespace, here.

gsmet replied:
We need this example to compile. We can't give a doc example that doesn't compile. So if you don't want to use the constant, just use `quarkus.log` as it was before. It is just an example so it doesn't need to be totally similar to the code we have in Quarkus.

sarxos replied:
Will change to "application.log" then. As I said before - I do not want to have `quarkus.log` file name because it's confusing with `quarkus.log` log namespace. By doing this it will compile and will remain non confusing :)

gsmet reviewed (changes requested):
I added a few comments. Could you check them out?

sarxos commented:
@gsmet I addressed some of your comments but I would like to discuss the remaining one. For a readability sake I would like to avoid having any mention of `quarkus.log` in this example in terms other than a log namespace. This is the core reason why I changed this asciidoc paragraph - because it was confusing as hell :)

gsmet commented on core/runtime/src/main/java/io/quarkus/runtime/logging/FileConfig.java:
```diff
@@ -40,15 +45,15 @@
 
     /**
-     * The file logging log level
+     * The name of file where logs will be stored.
```
I would go with `The name of the file in which logs will be written.`

sarxos replied:
Done.

cescoffier replied:
agreed, @sarxos can you change this one?

sarxos replied:
Hi @cescoffier This this already modified in accordance to @gsmet request. Please see https://github.com/quarkusio/quarkus/pull/2589/commits/aab71c4a73481ac56de7955b65cf863c65e040b7. These two commits will be squashed after I have final approval for these changes.

gsmet commented on docs/src/main/asciidoc/extension-authors-guide.adoc:
```diff
@@ -494,15 +494,15 @@ public class FileConfig {
 
     /**
-     * The file logging log level.
+     * The name of file where logs will be stored.
```
And we should apply the same change here.

sarxos replied:
Done.

cescoffier replied:
same comment here @sarxos

sarxos replied:
@cescoffier This this already modified in accordance to @gsmet request. Please see aab71c4. These two commits will be squashed after I have final approval for these changes.

gsmet reviewed (dismissed):
I added a couple more comments. Once you are done with them, I think it will be OK.

Could you squash everything in one commit at the end?

Thanks!

sarxos commented:
@gsmet I replaced `DEFAULT_LOG_FILE_NAME` with the `application.log`. Please let me know if this is ok and then I will squash these commits.

cescoffier commented:
Waiting for CI.

cescoffier commented:
@gsmet @dmlloyd are you ok with the latest changes?

dmlloyd commented:
To quote @gsmet:

> We need this example to compile. We can't give a doc example that doesn't compile. So if you don't want to use the constant, just use `quarkus.log` as it was before. It is just an example so it doesn't need to be totally similar to the code we have in Quarkus.

And round and round we go.  Using logging for the example was clearly a mistake!

sarxos commented:
Can we just leave it as it is now? In its current form it's sufficient to serve as an example, IMO.

cescoffier commented:
@sarxos yes we keep it, the idea of using log was taken a long time ago.

dmlloyd reviewed (commented):
I have no opinion one way or another about the change.

sarxos commented:
Hi @cescoffier,

What should I do with this pull request? Are we merging it or shall I close it as unnecessary? IMO it should be merged since existing example in its current form, together with related explanation, is really confusing. The changes I did are to clarify it, at least a little bit.

sarxos commented:
Commits are now squashed.

cescoffier commented:
I think we are good to go. This PR clearly shows an issue with our example (log) and we may need to come with something different - less controversial as an example.
